text/287-cli-deprecation-warnings.md

```diff
@@ -0,0 +1,190 @@
+I think it's worth it to talk about what solutions we _don't_ want,
+as that will strongly inform how we approach the problem.
+
+I think we can rule out any solution that requires parsing the customer's source code in search for usages of deprecated elements.
+The downsides of this approach include:
+
+* We would have to write a separate parser for each language supported by the CDK.
+* This additional parsing could have an adverse impact on the performance of CDK commands.
+
+If we agree that we don't want the above solution,
+then the only avenue we have left is injecting some extra code in the deprecated elements during compilation of the TypeScript code into JavaScript.
+There are two ways we can do that: through JSII, or purely through the TypeScript compiler.
+
+##### 2.1. Modifying JSII
+
+We could modify JSII to add a capability to it that allows injecting some code when compiling every deprecated element.
+It would involve writing a TypeScript transform,
+that adds some code whenever the deprecated element is accessed.
+Nick did a similar change in PR [#2348](https://github.com/aws/jsii/pull/2348)
+to add runtime version information to all JSII-compiled classes.
+
+Advantages:
+
+* Doing it in JSII wil make it easy to record the module the given element is in.
+
+Disadvantages:
+
+* Because the code adding the warning for the deprecated elements will be CDK-specific,
+  it will be pretty awkward to correctly pass what exactly that code should be from CDK to JSII
+  (and, additionally, that code will most likely be slightly different for each type --
+  class, method, property, etc. --
+  of deprecated element we support).
+* Requires coordination between the JSII and CDK projects
+  (merging a JSII PR, waiting for a release, updating CDK to use the new version, etc.),
+  which typically lengthens the total development time.
+
+##### 2.2. TypeScript decorators
+
+We can use [TypeScript decorators](https://www.typescriptlang.org/docs/handbook/decorators.html)
+to enhance the runtime JavaScript code with our custom logic.
+We can add an `awslint` rule that enforces that every element with the `@deprecated`
+JSDoc tag also needs to have the `@deprecated` decorator present.
+While decorators are still considered an experimental feature,
+our init templates and JSII enable them in `tsconfig.json`,
+which means the vast majority of our customers will have them enabled as well.
+
+Advantages:
+
+* Changes are needed only in the CDK project.
+
+Disadvantages:
+
+* It won't be possible to reliably get the module the deprecated element belongs to
+  (only its name).
+
+**My recommendation**: TypeScript decorators.
+
+### Is this a breaking change?
+
+No.
+
+### What are the drawbacks of this solution?
+
+The drawback of this solution is that we will overwhelm our customers with warnings,
+and train them to ignore the output of `cdk` commands because of that.
+
+### What alternative solutions did you consider?
```

Comment:
The alternative solution are described above. Move them here and only leave the proposed solution at the top

Reply:
Done.